kernel/libc/newlib/newlib_isatty.c

```diff
@@ -3,29 +3,43 @@
    newlib_isatty.c
    Copyright (C) 2004 Megan Potter
    Copyright (C) 2012 Lawrence Sebald
+   Copyright (C) 2024 Andress Barajas
 
 */
 
+#include <unistd.h>
+#include <errno.h>
+
+#include <sys/stat.h>
 #include <sys/reent.h>
 
+#define PTY_DEV (dev_t)('p' | ('t' << 8) | ('y' << 16))
+
 int isatty(int fd) {
-    /* Make sure that stdin, stdout, and stderr are shown as ttys, otherwise
-       they won't be set as line-buffered. */
-    if(fd >= 0 && fd <= 2) {
-        return 1;
+    struct stat statbuf;
+
+    if(fd < 0) {
+        errno = EBADF;
+        return 0;
     }
 
+    /* Make sure that stdin is shown as a tty, otherwise
+       it won't be set as line-buffered. */
+    if(fd == STDIN_FILENO)
+        return 1;
+
+    if(fstat(fd, &statbuf) == -1)
+        return 0;
+
+    /* Check if PTY */
+    if(statbuf.st_dev == PTY_DEV)
```

Comment:
```suggestion
    if(statbuf.st_dev == PTY_DEV)
```
Not gonna lie, like... This makes me kind of uncomfortable... In terms of design, the fact we're basically hard-coding this stuff as being TTY or not TTY by doing some intrusive inspection... and now `PTY_DEV` is something that isn't really part of the PTY interface and is a local private define rather than a public one that would at least keep the values synchronized.... 

I guess I'm fine doing this as part of this PR, but the *correct* solution really should be adding an `isatty()` virtual function/function pointer to the VFS interface itself so that each individual filesystem can uniformly report back whether it is to be treated as a TTY or not, and that logic remains encapsulated within each...

Also an alternative would be to add another flag to the statbuf or something...

Reply:
That makes sense.  I think it would be better to have that in another PR.  I think there are two open PRs currently that "cleanup" the fs.c file.  Maybe it can be done in one of those?  Im cool to create another PR to take the virtual function/function pointer approach after this one is merged.

Reply:
> Not gonna lie, like... This makes me kind of uncomfortable... In terms of design, the fact we're basically hard-coding this stuff as being TTY or not TTY by doing some intrusive inspection... and now `PTY_DEV` is something that isn't really part of the PTY interface and is a local private define rather than a public one that would at least keep the values synchronized....
> 
> I guess I'm fine doing this as part of this PR, but the _correct_ solution really should be adding a something like a `bool (*isatty)(file_t)` virtual function/function pointer to the VFS interface itself so that each individual filesystem can uniformly report back whether it is to be treated as a TTY or not, and that logic remains encapsulated within each...
> 
> Also an alternative would be to add another flag to the statbuf or something...
> 
> But like I said, probably can come later during a more clean-up/feature addition pass to the VFS, as I feel like there are other things that need to be added there as well.

If you look at how `isatty` is implemented in many platforms it often comes down to insane things like comparing to a list of known TTY driver device names. While I agree that `PTY_DEV` should stem directly from fs_pty, it seems silly to add a new virtual function that would only deviate/be used for a single driver. Unless it seems likely that we get other hardware-backed TTY implementations, but I really don't imagine that being the case.

Another flag in statbuf doesn't seem right either since that's so much part of the external stat system which normally has no such flags. Doing it at the vfs driver level as just a flag (rather than the virtual function that would basically return the value of such a flag) would be my preferred method, but even that seems like a lot for a singular exception that already has to have a lot of hardcoded stuff by standard.

Reply:
I just looked at several platforms (because I have no idea how it's implemented), and two of them that I just looked at for libc and Linux are calling into `tcgetattr()` which makes a call into `ioctl()`., which... we actually do have in the VFS.

I guess I agree with you. I forgot `ioctl()` is there as a kind of one-off single-command thing that is optional to implement on most filesystems and can be safely ignored for things that aren't supported, but that doesn't require a whole function pointer entry in the driver. 

Here's `isatty()` for glibc/POSIX:
https://codebrowser.dev/glibc/glibc/sysdeps/posix/isatty.c.html#27
Here's the underlying `tcgetattr()` function:
https://codebrowser.dev/glibc/glibc/sysdeps/unix/bsd/tcgetattr.c.html#__tcgetattr

Reply:
Nice find!  I'll take a look later today to use ioctl() instead

Reply:
I probably jumped the gun but I did the above implementation posted by @gyrovorbis.  Had to create more files. @QuzarDC let me know :^p. I think this is the only thing left blocking.